Support plugin/flags.vim file sourced on first flag access

dbarnett opened the pull request:
Partially fixes #189 (allowing plugins to start migrating to plugin/flags.vim but not discouraging the instant/ equivalent yet).

dbarnett committed:
Support plugin/flags.vim file sourced on first flag access

Partially fixes #189 (allowing plugins to start migrating to
plugin/flags.vim but not discouraging the instant/ equivalent yet).

dbarnett commented on autoload/maktaba/plugin.vim:
```diff
@@ -240,10 +242,10 @@ function! maktaba#plugin#Enter(file) abort
   endif
 
-  if l:filedir !=# 'autoload'
+  if l:filedir !=# 'autoload' && has_key(l:plugin.flags, l:filedir)
```
This change avoids triggering flags.vim to load where it's not needed and potentially wasteful/unsafe. Calling Flag could even try to load recursively if this maktaba#plugin#Enter was called from flags.vim.

dbarnett commented:
FWIW, this could add a bit of overhead trying to source a plugin/flags.vim file, but only in uncommon corner cases because most current plugins that will have any flags configured will already have `PLUGIN._loaded_flags` set via instant/flags.vim by the time that anything else would try to resolve flags.

Also this approach is a little simpler than what I described in #189 since instead of trying to be clever and queue up flags operations it just tries to ensure flags.vim is sourced before any flags operations are performed to pick up default values. This shouldn't preclude later improvements to do the clever queuing if we're able to get all the intricacies right.

malcolmr reviewed (approved):
If I understand this correctly:

- This is a no-op for any plugin with an `instant/flags.vim`; everything proceeds as normal.
- For any plugin without `instant/flags.vim`, this will trigger an extra load attempt if a caller (not Maktaba) tries to look up its flags (e.g. by using Glaive). But it's unlikely that there are many/any Maktaba plugins that both don't have any flags at all and have a caller that tries to look at their flags.
- This does assume that flags are only defined in `instant/flags.vim` specifically (previously they could have been in _any_ `instant/` file); I suspect that's a reasonable assumption though.

dbarnett committed:
Add comments and doc tweaks from code review

dbarnett commented on autoload/maktaba/plugin.vim:
```diff
@@ -180,6 +180,26 @@ function! s:FileHandleToFlagHandle(handle) abort
+  endif
+  try
+    call a:plugin.Load('flags')
+    if !a:plugin._loaded_flags
```
Added.

dbarnett commented on autoload/maktaba/plugin.vim:
```diff
@@ -225,6 +245,8 @@ function! maktaba#plugin#Enter(file) abort
   let l:plugin = maktaba#plugin#GetOrInstall(l:plugindir)
   let l:controller = l:plugin._entered[l:filedir]
+  let l:is_flags_file = (l:filedir ==# 'plugin' || l:filedir ==# 'instant')
+      \ && l:handle ==# 'flags^'
```
Added a comment. It's the weird "handle" syntax from SplitEnteredFile.

dbarnett commented on autoload/maktaba/plugin.vim:
```diff
@@ -240,7 +262,7 @@ function! maktaba#plugin#Enter(file) abort
   endif
 
-  if l:filedir !=# 'autoload'
+  if l:filedir !=# 'autoload' && !l:is_flags_file
```
Sure, added some comments to clarify all that.

dbarnett commented on autoload/maktaba/plugin.vim:
```diff
@@ -757,6 +783,8 @@ endfunction
 "   maktaba#plugin#Get('myplugin').flags.foo.Get()
 " <
+" except that the second version will not trigger flags to load if they haven't
+" been loaded already.
```
Right, I want to go a little further to discourage accessing `.flags`, possibly rename it to `._flags` and provide a `.Flags()` lookup method instead, but I don't want to get too verbose with usage contracts just yet in all the files I'd need to touch for that. The helpers ought to fully encapsulate having the flags resolved before letting you Set/Get them and I can't think of any reason to support poking at them partially initialized.

dbarnett commented on autoload/maktaba/plugin.vim:
```diff
@@ -506,10 +541,10 @@ function! s:CreatePluginObject(name, location, settings) abort
+    let l:plugin.flags.instant = maktaba#flags#Create('instant', {})
   endif
 
   " Load flags file first.
```
Actually I noticed this ApplySettings case probably also needs to EnsureFlagsLoaded first. I'm going to try to get some vroom coverage to repro an issue there along with adding the call to fix it.

dbarnett replied:
Alright, pushed 0382ab2 to add the missing corner case handling (in Setting.Apply) and test coverage.

dbarnett committed:
Ensure flags are loaded in more cases and add Plugin.AllFlags helper

Ensures that a plugin/flags.vim file gets loaded before HasFlags or
Setting.Apply calls and adds vroom test coverage for more of those
cases. Also defines a related AllFlags helper on the plugin object to
look up flags if needed and return them.

dbarnett commented on autoload/maktaba/plugin.vim:
```diff
@@ -757,6 +783,8 @@ endfunction
 "   maktaba#plugin#Get('myplugin').Flag('foo')
 "   maktaba#plugin#Get('myplugin').flags.foo.Get()
 " <
+" except that the second version will not trigger flags to load if they haven't
```
Sure, updated.

dbarnett commented on autoload/maktaba/plugin.vim:
```diff
@@ -757,6 +783,8 @@ endfunction
 "   maktaba#plugin#Get('myplugin').Flag('foo')
 "   maktaba#plugin#Get('myplugin').flags.foo.Get()
 " <
+" except that the second version will not trigger flags to load if they haven't
```
In a sense yes, but to actually break anything you'd need to update the plugins to move from instant/flags.vim to plugin/flags.vim.

dbarnett commented on autoload/maktaba/plugin.vim:
```diff
@@ -785,13 +813,16 @@ endfunction
 "   maktaba#plugin#Get('myplugin').Flag('foo', 'bar')
 "   maktaba#plugin#Get('myplugin').flags.foo.Set('bar')
 " <
+" except that the second version will not trigger flags to load if they haven't
```
Updated that example to not use `\.flags` so it doesn't need to call out that distinction.

dbarnett commented:
Ready for another look after various updates in 0382ab2 to add test coverage and fix ensuring flags file loaded in two cases I'd neglected (`Plugin.HasFlag` and `Setting.Apply`).

malcolmr reviewed (approved):
LGTM.

Nice, thanks!
